client/src/includes/tabs.js

```diff
@@ -0,0 +1,75 @@
+const DEFAULT_OPTIONS = {
+  tabActiveClass: 'active',
+  paneActiveClass: 'active',
+  tabLinkSelector: 'li > a',
+  initialActiveTab: 0,
+};
+
+class TabPanel {
+  constructor(el, options) {
+    this.el = el;
+    this.options = Object.assign({}, DEFAULT_OPTIONS, options);
+
+    this.setActiveTab = this.setActiveTab.bind(this);
+    this.render = this.render.bind(this);
+
+    this.tabs = Array.from(this.el.querySelectorAll(this.options.tabLinkSelector));
+    this.activeTab = this.tabs[this.options.initialActiveTab];
+
+    this.tabs.forEach(tab => tab.addEventListener('click', this.setActiveTab));
+
+    if (window.location.hash) {
+      // If there's a tab that matches the location hash, activate it
+      const newActiveTab = this.el.querySelector(`[href="${window.location.hash}"]`);
+      if (newActiveTab) {
+        // Simulate a click to activate tab. Triggers a render
+        newActiveTab.click();
+        return; // Early return to avoid a rerender
+      }
+    }
+
+    // Run render to ensure attributes are all properly set
+    this.render();
+  }
+
+  setActiveTab(evt) {
+    // If the clicked tab is already active, bail out
+    if (this.activeTab === evt.target) return;
+
+    this.activeTab = evt.target;
+    window.history.replaceState(null, null, this.activeTab.href);
+    this.render();
+    evt.stopPropagation();
+    evt.preventDefault();
+  }
+
+  render() {
+    this.tabs.forEach(tab => {
+      let pane;
+      if (tab === this.activeTab) {
+        // Add active class and ARIA attribute to active tab
+        tab.classList.add(this.options.tabActiveClass);
+        tab.setAttribute('aria-selected', 'true');
+        // Show active pane
+        pane = document.getElementById(tab.getAttribute('aria-controls'));
+        pane.classList.add(this.options.paneActiveClass);
+        pane.hidden = false;
```

Comment:
Yes. I have no idea why certain attributes are exposed bare to the DOM and others aren't, but https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/hidden

Looks like it might not have IE (pre-Edge) support though, if that's a concern. The CSS classes should provide an adequate fallback, though I haven't tested.

Reply:
👍 I’ll check.